src/gui/editors/SongEditor.cpp

```diff
@@ -52,26 +52,111 @@
 #include "PianoRoll.h"
 #include "Track.h"
 
-positionLine::positionLine( QWidget * parent ) :
-	QWidget( parent )
+const QVector<double> positionLine::m_zoomLevels =
+		{ 0.125f, 0.25f, 0.5f, 1.0f, 2.0f, 4.0f, 8.0f, 16.0f };
+
+positionLine::positionLine( QWidget * parent, ComboBoxModel * zoom, Song * song ) :
+	QWidget( parent ),
+	m_tailGradient ( false ),
+	m_lineColor (0, 0, 0, 0)
 {
-	setFixedWidth( 1 );
+	m_x = 0;
+	m_y = 0;
+	m_width = 8;
+	currentZoom = zoom;
+	p_song = song;
+	resize( m_width, height() );
+	
 	setAttribute( Qt::WA_NoSystemBackground, true );
+	setAttribute( Qt::WA_TransparentForMouseEvents );
 }
 
-
-
-
 void positionLine::paintEvent( QPaintEvent * pe )
 {
 	QPainter p( this );
-	p.fillRect( rect(), QColor( 255, 255, 255, 153 ) );
+	
+	// Resize based on the zoom value
+	m_width = 8.0f * m_zoomLevels[ currentZoom->value() ];
+	resize( m_width, height() );
+	
+	// If width is 1, we don't need a gradient
+	if (m_width == 1)
+	{
+		p.fillRect( rect(),
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 153) );
+	}
+	
+	// If width > 1, we need the gradient
+	else
+	{
+		// Create the gradient trail behind the line
+		QLinearGradient gradient( rect().bottomLeft(), rect().bottomRight() );
+		
+		// If gradient is enabled, we're in focus and we're playing, enable gradient
+		if (p_song->isPlaying() && m_tailGradient && 
+			p_song->playMode() == Song::Mode_PlaySong)
+		{
+			gradient.setColorAt((double)( ( width() - 1.0f )/width() ),
+				QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 60) );
+		}
+		else
+		{
+			gradient.setColorAt((double)( ( width() - 1.0f )/width() ),
+				QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 0) );
+		}
+		
+		// Fill in the remaining parts
+		gradient.setColorAt(0,
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 0) );
+		gradient.setColorAt(1,
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 153) );
+		
+		// Fill line
+		p.fillRect( rect(), gradient );
+	}
+}
+
+int positionLine::width( void )
+{
+	m_width = 8.0f * m_zoomLevels[ currentZoom->value() ];
+	return m_width;
 }
 
+// synonymous to move(), but use this instead of move() so position line can log changes in its m_x & m_y
+void positionLine::go( int x, int y )
+{
+	move( x, y );
+	m_x = x + width() - 1;
+	m_y = y;
+}
+
+// update width if zoom changes
+void positionLine::zoomUpdate()
+{ move ( m_x - width() + 1, m_y ); }
+
+// Essentially repaints so that gradient does not persist when stopped (this is a bug fix)
+void positionLine::playStateChanged()
+{ repaint(); }
```

Comment:
Prefer `update` to `repaint`: https://doc.qt.io/qt-5/qwidget.html#repaint.

Reply:
All `repaint()`s were changed to `update()`s